File: internal/handlers/join_waitlist.go
package handlers

import (
	"encoding/json"
	"log"
	"net/http"
	"os"

	"github.com/labstack/echo/v4"
	"github.com/sendgrid/sendgrid-go"
)

type Contact struct {
	Firstname string `json:"firstname"`
	Lastname  string `json:"lastname"`
	Email     string `json:"email"`
}

type ContactRequest struct {
	ListIds  []string  `json:"list_ids"`
	Contacts []Contact `json:"contacts"`
}

func (h *Handler) JoinWaitlist(c echo.Context) error {

	// Simple validation
	if len(c.FormValue("firstname")) == 0 {
		c.NoContent(400)
	}

	if len(c.FormValue("lastname")) == 0 {
		c.NoContent(400)
	}

	if len(c.FormValue("email")) == 0 {
		c.NoContent(400)
	}

	sendGridApiKey := os.Getenv("SENDGRID_API_KEY")
	sendGridListId := os.Getenv("SENDGRID_LIST_ID")

	data := ContactRequest{
		ListIds: []string{sendGridListId},
		Contacts: []Contact{
			{
				Firstname: c.FormValue("firstname"),
				Lastname:  c.FormValue("lastname"),
				Email:     c.FormValue("email"),
			},
		},
	}

	// fmt.Println(data)

	jsonBytes, err := json.Marshal(data)
	if err != nil {
		log.Println("failed to json marshal contacts")
		return c.NoContent(http.StatusInternalServerError)
	}

	sendGridUrl := "https://api.sendgrid.com"
	request := sendgrid.GetRequest(sendGridApiKey, "/v3/marketing/contacts", sendGridUrl)
	request.Method = "PUT"
	request.Body = jsonBytes

	response, err := sendgrid.API(request)
	if err != nil {
		log.Println("Sendgrid api error")
		return c.NoContent(http.StatusInternalServerError)
	}

	if response.StatusCode != 202 {
		log.Println(response.Body)
		return c.NoContent(http.StatusInternalServerError)
	}

	log.Println(response.Body)
	log.Println("Contact saved!!")
	return c.NoContent(http.StatusOK)
}

File: web/routes/join_waitlist.py
from fastapi import APIRouter, Response, Form
from typing import Annotated
import os
from sendgrid import SendGridAPIClient
from utils.logger import log

router = APIRouter()

@router.post("/join-waitlist")
async def join_waitlist(
  firstname: Annotated[str, Form()],
  lastname: Annotated[str, Form()],
  email: Annotated[str, Form()]
) -> Response:
  # print(f"Received join waitlist request for {firstname} {lastname} ({email})")
  sendgrid_api_key = os.getenv("SENDGRID_API_KEY")
  sendgrid_list_id = os.getenv("SENDGRID_LIST_ID")

  sg = SendGridAPIClient(api_key=sendgrid_api_key)

  data = {
    "list_ids": [sendgrid_list_id],
    "contacts": [{ "email": email, "first_name": firstname, "last_name": lastname}]
  }

  try:response = sg.client.marketing.contacts.put(request_body=data) # type: ignore
  except Exception as e:
    log.error(e)
    return Response(status_code=500)

  log.info(response.status_code, response.body)
  return Response(status_code=200)
